packages/clerk-js/src/ui/common/withRedirect.tsx

```diff
@@ -0,0 +1,98 @@
+import { useClerk } from '@clerk/shared/react';
+import type { Clerk, ClerkOptions, EnvironmentResource } from '@clerk/types';
+import type { ComponentType } from 'react';
+import React from 'react';
+
+import { warnings } from '../../core/warnings';
+import type { ComponentGuard } from '../../utils';
+import { sessionExistsAndSingleSessionModeEnabled } from '../../utils';
+import { useEnvironment, useOptions, useSignInContext, useSignUpContext } from '../contexts';
+import { useRouter } from '../router';
+import type { AvailableComponentProps } from '../types';
+
+type RedirectUrl = (opts: { clerk: Clerk; environment: EnvironmentResource; options: ClerkOptions }) => string;
+
+export function withRedirect<P extends AvailableComponentProps>(
+  Component: ComponentType<P>,
+  condition: ComponentGuard,
+  redirectUrl: RedirectUrl,
+  warning?: string,
+): (props: P) => null | JSX.Element {
+  const displayName = Component.displayName || Component.name || 'Component';
+  Component.displayName = displayName;
+
+  const HOC = (props: P) => {
+    const { navigate } = useRouter();
+    const clerk = useClerk();
+    const environment = useEnvironment();
+    const options = useOptions();
+
+    const shouldRedirect = condition(clerk, environment, options);
+    React.useEffect(() => {
+      if (shouldRedirect) {
+        if (warning && environment.displayConfig.instanceEnvironmentType === 'development') {
```

Comment:
🔧 Let's introduce a `isDevelopmentFromPK(publishableKey)`  in `@clerk/shared/keys` and use that instead of displayConfig.